app-marketplace-v2-spec/architecture.md

```diff
@@ -0,0 +1,146 @@
+# Architecture.
+
+  * [Cloud Apps.](#cloud-apps)
+    + [Authentication with Mattermost.](#authentication-with-mattermost)
+    + [Authentication with Upstream Applications.](#authentication-with-upstream-applications)
+    + [Cloud App to Mattermost.](#cloud-app-to-mattermost)
+    + [Mattermost to Cloud App.](#mattermost-to-cloud-app)
+    + [Functions](#functions)
+    + [Interactive Inputs.](#interactive-inputs)
+      - [Slash Commands](#slash-commands)
+      - [Widgets](#widgets)
+      - [Dialogs](#dialogs)
+    + [Change notifications.](#change-notifications)
+    + [Autocomplete lookups](#autocomplete-lookups)
+    + [Lifecycle - Install](#lifecycle---install)
+    + [Lifecycle Uninstall.](#lifecycle-uninstall)
+    + [Configuration.](#configuration)
+    + [mattermost-cloud-app.json](#mattermost-cloud-appjson)
+  * [mattermost-plugin-cloud-apps](#mattermost-plugin-cloud-apps)
+  * [mattermost-server](#mattermost-server)
+    + [Minimal Changes to the Core Code.](#minimal-changes-to-the-core-code)
+  * [Cloud App Marketplace](#cloud-app-marketplace)
+
+## Cloud Apps.
```

Comment:
Do we consider plugins, custom slash commands, webhooks and bot accounts as Apps or are those separate entities?